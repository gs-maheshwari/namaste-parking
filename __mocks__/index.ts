import { VehicleType } from "@/lib";
import { ParkingSession, ParkingSpace } from "@/types";

export const mockParkingSpaces: ParkingSpace[] = [
  {
    parkingSpaceId: 1,
    isOccupied: false,
    occupancy: 1,
    capacity: 50,
    vehicleType: null,
  },
  {
    parkingSpaceId: 2,
    isOccupied: false,
    occupancy: 8,
    capacity: 80,
    vehicleType: VehicleType.CAR,
  },
  {
    parkingSpaceId: 3,
    isOccupied: false,
    occupancy: 16,
    capacity: 20,
    vehicleType: VehicleType.MOTOR,
  },
];

export const mockParkingSessions = {
  parkingSessions: [
    {
      parkingSessionId: "8ad85151-7da3-4a9b-91fe-0bbcd9279b30",
      parkingSpaceId: 2,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 6,
      sessionStartedAt: "2024-11-18T10:21:07.738Z",
      sessionEndedAt: "2024-11-18T10:27:17.623Z",
      vehicleLicensePlate: "FG-7863",
      vehicleType: "CAR",
    },
    {
      parkingSessionId: "ef026d47-70ce-45f6-9887-26135ef8768f",
      parkingSpaceId: 2,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 0,
      sessionStartedAt: "2024-11-15T21:04:44.302Z",
      sessionEndedAt: "2024-11-15T21:05:11.770Z",
      vehicleLicensePlate: "PF-1324",
      vehicleType: "CAR",
    },
    {
      parkingSessionId: "14737b70-3918-47d6-a09c-2497f0c3e608",
      parkingSpaceId: 3,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 0,
      sessionStartedAt: "2024-11-13T16:14:26.081Z",
      sessionEndedAt: "2024-11-13T16:14:49.609Z",
      vehicleLicensePlate: "PF-121296",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "53a1fa08-49ae-4c02-99fd-c1923d07e2d6",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 6856,
      sessionStartedAt: "2024-11-13T16:09:35.157Z",
      sessionEndedAt: "2024-11-18T10:26:02.016Z",
      vehicleLicensePlate: "GH-9823",
      vehicleType: "CAR",
    },
    {
      parkingSessionId: "21dfcfa4-6f2d-4bab-bad7-4ee557fedc1e",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 2560,
      sessionStartedAt: "2024-11-11T21:29:13.833Z",
      sessionEndedAt: "2024-11-13T16:09:19.288Z",
      vehicleLicensePlate: "KL-9881",
      vehicleType: "CAR",
    },
    {
      parkingSessionId: "1db3ddea-7537-4bf0-991e-ee40bdf74c01",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 120,
      sessionStartedAt: "2024-11-11T21:29:07.170Z",
      sessionEndedAt: "2024-11-11T23:29:26.742Z",
      vehicleLicensePlate: "KL-9322",
      vehicleType: "CAR",
    },
    {
      parkingSessionId: "8bc4f4dc-a725-44ba-8f2d-e43a82f3c7cf",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:29:02.201Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "WA-9322",
      vehicleType: "CAR",
    },
    {
      parkingSessionId: "a8f77a22-ec47-45ac-a98a-215abe2d0e0c",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 117246,
      sessionStartedAt: "2024-11-11T21:28:48.506Z",
      sessionEndedAt: "2025-02-01T07:34:40.164Z",
      vehicleLicensePlate: "QA-9322",
      vehicleType: "CAR",
    },
    {
      parkingSessionId: "d79902b4-8baf-4396-acd1-c09cb4dbd73c",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 2586,
      sessionStartedAt: "2024-11-11T21:28:41.502Z",
      sessionEndedAt: "2024-11-13T16:34:35.720Z",
      vehicleLicensePlate: "RA-9322",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "96203099-bc7e-43db-9f3f-c8e212689e22",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:28:35.626Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "ZA-9322",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "97241c6d-39ac-457b-81df-59bcf8f3775f",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 9418,
      sessionStartedAt: "2024-11-11T21:28:29.646Z",
      sessionEndedAt: "2024-11-18T10:26:53.907Z",
      vehicleLicensePlate: "ZA-9911",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "3ad9cc4c-0c53-40b8-95ec-f8e17b6b1638",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:28:21.880Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "ZA-9712",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "c7dec29a-ab4e-4a05-a477-942d24246173",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:28:14.234Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "ZA-8719",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "aefbae5f-932e-4414-9afe-d972a4742fee",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 9419,
      sessionStartedAt: "2024-11-11T21:28:04.286Z",
      sessionEndedAt: "2024-11-18T10:27:12.244Z",
      vehicleLicensePlate: "OP-8179",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "8a2d1550-1e5e-4acc-af32-b542d03aa09f",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:27:59.486Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "KL-8179",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "fb5f93c4-490e-4138-af2b-65e282dc5163",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 117250,
      sessionStartedAt: "2024-11-11T21:27:53.790Z",
      sessionEndedAt: "2025-02-01T07:37:26.773Z",
      vehicleLicensePlate: "QW-8179",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "651a3135-50f0-4c2a-af36-16810d9c2b6a",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:27:46.338Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "KL-8179",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "f4074553-1cdc-4266-8575-86dff27c9fdc",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:27:41.474Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "GS-8179",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "3c08a199-ebb6-4c68-b85c-62405f3f6198",
      parkingSpaceId: 1,
      isSessionEnded: true,
      sessionLengthInHoursMinutes: 117437,
      sessionStartedAt: "2024-11-11T21:27:35.900Z",
      sessionEndedAt: "2025-02-01T10:44:34.740Z",
      vehicleLicensePlate: "DC-8179",
      vehicleType: "MOTOR",
    },
    {
      parkingSessionId: "35c61359-fcc3-497a-8c91-8dfe3307ce10",
      parkingSpaceId: 1,
      isSessionEnded: false,
      sessionLengthInHoursMinutes: null,
      sessionStartedAt: "2024-11-11T21:27:28.594Z",
      sessionEndedAt: null,
      vehicleLicensePlate: "DC-8171",
      vehicleType: "CAR",
    },
  ],
  parkingSessionsTotalCount: 524,
};
