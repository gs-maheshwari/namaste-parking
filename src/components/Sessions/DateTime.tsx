"use client";

import { useEffect, useState } from "react";

const formatDate = (isoString: string) => {
  return new Intl.DateTimeFormat(navigator.language, {
    dateStyle: "medium",
    timeStyle: "short",
  }).format(new Date(isoString));
};

export const DateTime = ({ isoDate }: { isoDate: string }) => {
  const [formattedTime, setFormattedTime] = useState("Loading...");

  useEffect(() => {
    setFormattedTime(formatDate(isoDate));
  }, [isoDate]);

  return <span>{formattedTime}</span>;
};
