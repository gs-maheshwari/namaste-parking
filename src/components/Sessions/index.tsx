"use client"

import { useMemo, useCallback, useState } from "react"
import useSWR from "swr"
import type { ParkingSession, ParkingSessionsResponse } from "@/types"
import { endParkingSession, getParkingSessions } from "@/services"
import { VehicleType } from "@/lib"

import { Button, Table, TableFilter } from "../ui"
import { DateTime } from "./DateTime"
import Pagination from "../Pagination"

const Sessions = ({ sessionsData }: { sessionsData: ParkingSessionsResponse }) => {
  const [filters, setFilters] = useState<Record<string, string>>({});
  const [endingSessionIds, setEndingSessionIds] = useState<Record<string, boolean>>({})

  const { data, mutate } = useSWR<ParkingSessionsResponse>(
    [filters],
    ([filters]) => getParkingSessions(new URLSearchParams(
      Object.entries(filters)
    ).toString()),
    {
      fallbackData: sessionsData,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
    },
  )

  const onEndSession = useCallback(
    async (sessionId: string) => {
      setEndingSessionIds((prev) => ({ ...prev, [sessionId]: true }));
      await endParkingSession(sessionId);
      mutate();
      setEndingSessionIds((prev) => ({ ...prev, [sessionId]: false }));
    },
    [mutate],
  )

  const columns = useMemo(
    () => [
      { key: "parkingSessionId", header: "ID" },
      { key: "parkingSpaceId", header: "Space ID" },
      {
        key: "status",
        header: "Status",
        render: (session: ParkingSession) => (session.isSessionEnded ? "Inactive" : "Active"),
      },
      {
        key: "sessionLengthInHoursMinutes",
        header: "Length",
        render: (session: ParkingSession) => `${convertToHoursMinutes(session.sessionLengthInHoursMinutes || 0)}`,
      },
      {
        key: "sessionStartedAt",
        header: "Start Time",
        render: (session: ParkingSession) => <DateTime isoDate={session.sessionStartedAt} />,
      },
      {
        key: "sessionEndedAt",
        header: "End Time",
        render: (session: ParkingSession) =>
          session.sessionEndedAt ? <DateTime isoDate={session.sessionEndedAt} /> : "N/A",
      },
      { key: "vehicleLicensePlate", header: "License Plate" },
      { key: "vehicleType", header: "Vehicle Type" },
      {
        key: "revenue",
        header: "Revenue",
        render: (session: ParkingSession) =>
          session.sessionEndedAt && session.parkingSpaceId !== 1
            ? `â‚¬ ${getRevenue(session.sessionLengthInHoursMinutes || 0, session.vehicleType)}`
            : "N/A",
      },
      {
        key: "actions",
        header: "Actions",
        render: (session: ParkingSession) =>
          !session.isSessionEnded && (
            <Button onClick={() => onEndSession(session.parkingSessionId)} disabled={endingSessionIds[session.parkingSessionId]} variant="primary">
              {endingSessionIds[session.parkingSessionId] ? "Ending..." : "End"}
            </Button>
          ),
      },
    ],
    [onEndSession, endingSessionIds],
  )

  const filterConfig = [
    {
      key: "vehicleType",
      type: "select" as const,
      label: "Vehicle Type",
      options: [
        { value: VehicleType.CAR, label: "CAR" },
        { value: VehicleType.MOTOR, label: "MOTOR" },
      ],
    },
    { key: "sessionStartedAtFrom", type: "date" as const, label: "Start Date" },
    { key: "sessionEndedAtTo", type: "date" as const, label: "End Date" },
    {
      key: "isSessionEnded",
      type: "select" as const,
      label: "Status",
      options: [
        { value: "false", label: "Active" },
        { value: "true", label: "Inactive" },
      ],
    },
  ]

  const onFilterChange = useCallback((newFilters: Record<string, string>) => {
    setFilters(newFilters)
  }, [])

  return (
    <>
      <TableFilter filters={filters} filterConfig={filterConfig} onFilterChange={onFilterChange} />
      {data?.parkingSessions && <Table data={data.parkingSessions} columns={columns} />}
      {data?.parkingSessionsTotalCount || 0 > 0 ? (
        <Pagination
          totalItems={data?.parkingSessionsTotalCount || 0}
          onPageChange={(newOffset, limit) =>
            setFilters((prev) => ({
              ...prev,
              offset: newOffset.toString(),
              limit: limit.toString(),
            }))
          }
        />
      ) : null}
    </>
  )
}

const convertToHoursMinutes = (totalMinutes: number) => {
  const hours = Math.floor(totalMinutes / 60)
  const minutes = totalMinutes % 60
  return `${hours}h ${minutes}m`
}

const getRevenue = (totalMinutes: number, vehicleType: VehicleType) => {
  const hours = Math.ceil(totalMinutes / 60)
  const revenue = vehicleType === VehicleType.CAR ? hours * 5 : hours * 3
  return revenue
}

export default Sessions

