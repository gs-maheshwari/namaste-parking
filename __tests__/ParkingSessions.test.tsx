import { render } from "@testing-library/react";
import Sessions from "@/components/Sessions";
import { mockParkingSessions } from "../__mocks__";

describe("Sessions Component", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = render(
      <Sessions
        sessions={mockParkingSessions.parkingSessions}
        totalCount={mockParkingSessions.parkingSessionsTotalCount}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
