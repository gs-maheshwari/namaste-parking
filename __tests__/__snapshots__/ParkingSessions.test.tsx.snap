// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sessions Component renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="filters"
  >
    <div
      class="filterItem"
    >
      <label
        for="vehicleType"
      >
        Vehicle Type
      </label>
      <select
        id="vehicleType"
      >
        <option
          value=""
        >
          All
        </option>
        <option
          value="CAR"
        >
          CAR
        </option>
        <option
          value="MOTOR"
        >
          MOTOR
        </option>
      </select>
    </div>
    <div
      class="filterItem"
    >
      <label
        for="sessionStartedAtFrom"
      >
        Start Date
      </label>
      <input
        id="sessionStartedAtFrom"
        type="date"
        value=""
      />
    </div>
    <div
      class="filterItem"
    >
      <label
        for="sessionEndedAtTo"
      >
        End Date
      </label>
      <input
        id="sessionEndedAtTo"
        type="date"
        value=""
      />
    </div>
    <div
      class="filterItem"
    >
      <label
        for="isSessionEnded"
      >
        Status
      </label>
      <select
        id="isSessionEnded"
      >
        <option
          value=""
        >
          All
        </option>
        <option
          value="false"
        >
          Active
        </option>
        <option
          value="true"
        >
          Inactive
        </option>
      </select>
    </div>
  </div>
</div>
`;
